### ISSUE TITLE
[ENH]: Adding jitter mode in `scatter`

### ISSUE THREAD

**Issue by edoaltamura** [matplotlib/matplotlib#27935]
### Problem

Is there a plan to add a 'jitter' functionality to `scatter`, so that markers are re-positioned to avoid overlap? An example of this idea is described [here](https://datavizproject.com/data-type/jitter-plot/).

### Proposed solution

Provided that the number of overlapping markers isn't too large, the re-positioning can be achieved by minimizing a measure of distance over the shape centers (for circles). A basic implementation could be based on [this discussion](https://stackoverflow.com/questions/46131572/making-a-non-overlapping-bubble-chart).

**Comment by tacaswell** [comment]
This has come up and was rejected at least once before: https://github.com/matplotlib/matplotlib/issues/2750

If we did take some version of this functionality, it should be its own function/method rather than being tacked onto scatter.  This really only makes sense in the narrow case of one of x being categorical so we should let it only work for catagorical data (or at least require integers or maybe it takes a scalar x and a list of y or a sequence x and a sequence of sequence of y (e.g. the groupby) or match the API of boxplot and violinplot).  

I also suspect that there is going to be a large number of knobs people will want to have added to control how the jitter works which should not be shoe-horned into (already too complicated) signature of `scatter`.  

I suspect that a majority of these cases are with the realm of what seaborn targets and in that case, we should point people at `sns.swarmplot` (https://seaborn.pydata.org/generated/seaborn.swarmplot.html ).

This sort of thing could be made easier by the work in https://github.com/matplotlib/data-prototype where this could be implemented as step in the processing pipeline.

That said, I wonder if this can be achieved by writing a very custom Transform (https://matplotlib.org/stable/api/transformations.html#matplotlib.transforms.Transform) that is offered the full set of x, y and returns the "jittered" x, y.  If you write that, then I think you can do `ax.scatter(..., offset_transform=JitterTransfrom())` and it will "just work".

------

In summary my views on the ways this can be addressed:

 - adding API to `scatter` to jitter: hard no
 - adding a new method that pull swarmplot upstream: maybe leaning no
 - adding a `Transform` subclass that does the jitter + an example: maybe, leaning yes
 - new third-party library that implements hive/swarm/jitter plots: you don't need our permission, but why not just use seaborn?

**Comment by edoaltamura** [comment]
Thanks for the response @tacaswell, that's useful and clearly articulated. I like the `Transform` strategy to implement the jitter. Would you see value in exploring this?

**Comment by jklymak** [comment]
First, hard agree this cannot be tacked onto `scatter`.   It's far more akin to violin or box plots, and the x-value needs to be categorical since you want to use "jitter" in x to move the points around. 

>  I like the Transform strategy to implement the jitter. Would you see value in exploring this?

I think that is going to be _very_ hard as you are going to want the jitter in screen space, information of which we don't naturally pipe down to the the Transform.  Not impossible, but it's not going to be straightforward.  I think you'd be better to write a wrapper and not try and overload `scatter`.

**Comment by tacaswell** [comment]
We have transforms in the annotation context that do absolute offsets so it can be done.  I agree not easy, but I think it will be the fun kind of hard.

**Comment by scottshambaugh** [comment]
Throwing my +1 for some sort of native jittering, it's something I go to reach for fairly often and rediscover that it's not implemented. Was reading https://nightingaledvs.com/ive-stopped-using-box-plots-should-you/ this morning and am coming around to seeing it as a fairly fundamental data viz.

**Comment by jklymak** [comment]
I'm fine with an example or even a method - I just don't think the API should be an option to `scatter`.  

As for boxplots, it is hard to come up with a reason why you would ever want to use those, but I'm not convinced "jitter" plots are the right way to go either.  Multiple histograms, or 2-D histograms seem much better than a weird extra dimension where you are to artistically pick the density out by eye.

**Comment by scottshambaugh** [comment]
Scatter plots already handle categorical data seamlessly, but I agree with @tacaswell that we should break out a function to do this specifically, so that jitter can't be applied to non-categorical data (going so far as to raise an error? maybe not, as some categories might be numeric?). Note also that *both* axes might be categorical, in which case we will want to apply jitter in both directions.

**Comment by scottshambaugh** [comment]
Seaborn's `stripplot` API, for reference: https://seaborn.pydata.org/generated/seaborn.stripplot.html#seaborn.stripplot
